[danxuZhang] iP

imdxz opened the pull request.

Commit:
Add Increment: Level-0

Implement an initial skeletal version of the Duke that simply greets
the user and exits.

Commit:
Add helper method to print divider

Commit:
Add Level-1: Greet, Echo, Exit

Implement a skeletal version of Duke that starts by greeting the user,
simply echos commands entered by the user, and exits when the user types bye.

Change indentation from one space to one tab.

Commit:
Add Level-2: Add, List

Add the ability to store whatever text entered by the user and
display them back to the user when requested.

Commit:
Reformat Prompt Messages

Commit:
Fix typo

Commit:
Add Level-3: Mark as Done And A-Classes

Add the ability to mark tasks as done. Add the ability to change the status back to not done.

Use a Task class to represent tasks.

Commit:
Reformat code and add comments

Commit:
Rename marking status to avoid confusions

Commit:
Add Todo class

Add Todo class to represents tasks without any date/time attached to it.

Modify Duky main program to be open for adding more task types.

Commit:
Change accessibility of attributes in Task class.

Commit:
Add Deadline class

Commit:
Add Event class

Commit:
Reformat code

Commit:
Fix mark/unmark bug

When the user inputs command such as mark paper, the program will get an
exception. Now the program can handle situations when the user enters
mark/unmark with non-numeric values.

Commit:
Add text-ui-test

Commit:
Remove unnecessary whitespaces.

Commit:
Format code and add comments.

JuneNita commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,229 @@
+        assert cmdArgs[0].equals("deadline");
+        StringBuilder content = new StringBuilder();
+        StringBuilder by = new StringBuilder();
+        boolean buildContent = true;
```
I think this boolean variable (buildContent) could start with is, has, was, etc to make it more intuitive

JuneNita replied:
*only line 140

JuneNita commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,229 @@
+            return createEvent(cmdArgs);
+        default:
+            throw new IllegalArgumentException("Invalid task type!");
+        }
```
Nice and organized layout for switch statement

JuneNita commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,95 @@
+public class Task {
+
+    protected String content;
+    protected boolean isMarked;
```
Nice job with naming the variables 👍

JuneNita commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,229 @@
+import java.util.ArrayList;
+import java.util.Scanner;
```
I like how you list the imported class explicitly, good job

JuneNita commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,229 @@
+     * @param cmd user command.
+     * @return a task instance (can be event, deadline, etc).
+     * @throws IllegalArgumentException when user input an invalid command
+     */
```
I like how you comment the code following Javadoc style 👍

Commit:
Add getters and setters

Commit:
Reformat constructions of Task related classes.

Currently when creating `Deadline`, `Event` and `Todo classes`, functions that
are used to process user input commands like `createDeadline()`,
`createEvent()` are called in the `Duck.class`.

After modification, these parts are moved to the constructors of the
respective classes.

The implementation of creating these three classes from input arguments
are also modified to use index of `by` or `from`, `to`, making the code
logic easier to read.

The name of the personal assisstant is changed to Tohru.

Text-UI-Test is updated.

Commit:
Add Level 5: Handle Errors

Commit:
Merge branch 'branch-Level-5'

Commit:
Add Increment: A-Packages

Organize the classes into suitable java packages.

Commit:
Merge branch 'branch-A-Packages'

Commit:
Add Level 6: Delete

Add support for deleting tasks from the list.

Commit:
Add Level 7: Save

Save the tasks in the hard disk automatically whenever the task list
changes. Load the data from the hard disk when Duke starts up.

Commit:
Merge branch 'branch-Level-6'

Commit:
Merge branch 'branch-Level-7'

Commit:
Update .gitignore

Commit:
Reformat welcome message

theopin commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,104 @@
+        this.to = to;
+    }
+
+    public String toCSV() {
+        return "event," +
+                super.toCSV() + "," +
+                "\"" + from + "\"" + "," +
+                "\"" + to + "\"";
```
Seems to be a heavy usage of magic literals (string in this case). Do consider refactoring them into named constants to enhance readability of your code

theopin commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,104 @@
+     * @throws IllegalArgumentException exceptions with message when (part of) input is missing.
+     */
+    public Event(String[] args) throws IllegalArgumentException {
+        assert args[0].equals("event");
+        StringBuilder content = new StringBuilder();
+        StringBuilder from = new StringBuilder();
+        StringBuilder to = new StringBuilder();
+
+        int fromIndex = -1;
+        int toIndex = -1;
+        for (int i = 0; i < args.length; ++i) {
+            if (args[i].equals("/from")) {
+                fromIndex = i;
+            } else if (args[i].equals("/to")) {
+                toIndex = i;
+            }
+        }
+
+        if (fromIndex == -1 || toIndex == -1) {
+            throw new IllegalArgumentException("☹ OOPS!!! Cannot find from or to time of the event!");
+        }
+
+        for (int i = 1; i < args.length; ++i) {
+            if (i < fromIndex) {
+                content.append(args[i]).append(" ");
+            } else if (i > fromIndex && i < toIndex) {
+                from.append(args[i]).append(" ");
+            } else if (i > toIndex) {
+                to.append(args[i]).append(" ");
+            }
+        }
+
+        if (content.toString().isEmpty() || from.toString().isEmpty() || to.toString().isEmpty()) {
+            throw new IllegalArgumentException("☹ OOPS!!! The content/from/to of an event cannot be empty!");
+        }
+
+        this.content = content.toString().trim();
+        this.from = from.toString().trim();
+        this.to = to.toString().trim();
+    }
```
Seems to be a long constructor. Do try identifying common behaviour within it and extracting them as a separate method to enhance readability and maintainability

theopin commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,277 @@
+                + "|____/ \\__,_|_|\\_\\___|\n";
+        System.out.println("Hello from\n" + logo);
+
+        try {
+            readData(DATA_PATH + "/" + DATA_FILE);
+            System.out.println(spacer + "Read data from file " + DATA_PATH + "/" + DATA_FILE);
+        } catch (IOException ioe) {
+            System.out.println(spacer + "Create new data file in " + DATA_PATH + "/" + DATA_FILE);
+            initData(DATA_PATH, DATA_FILE);
+        }
+        // greeting messages
+        printDivider();
```
Missing a finally clause, that includes code that runs regardless of outcome of try catch (in this case line 245 to end of method)

theopin commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,277 @@
+    }
+
+    public static void main(String[] args) {
+        String logo = " ____        _        \n"
+                + "|  _ \\ _   _| | _____ \n"
+                + "| | | | | | | |/ / _ \\\n"
+                + "| |_| | |_| |   <  __/\n"
+                + "|____/ \\__,_|_|\\_\\___|\n";
+        System.out.println("Hello from\n" + logo);
```
Code chunks such as this that resemble a particular behaviour can be abstracted into a single method call to maintain a similar level of abstraction across the method (in this case, your switch statements). This will help enhance readability of your code.

theopin commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,277 @@
+            String line = fileScanner.nextLine();
+            String otherThanQuote = " [^\"] ";
+            String quotedString = String.format(" \" %s* \" ", otherThanQuote);
+            String regex = String.format("(?x) " + // enable comments, ignore white spaces
+                            ",                         " + // match a comma
+                            "(?=                       " + // start positive look ahead
+                            "  (?:                     " + //   start non-capturing group 1
+                            "    %s*                   " + //     match 'otherThanQuote' zero or more times
+                            "    %s                    " + //     match 'quotedString'
+                            "  )*                      " + //   end group 1 and repeat it zero or more times
+                            "  %s*                     " + //   match 'otherThanQuote'
+                            "  $                       " + // match the end of the string
+                            ")                         ", // stop positive look ahead
```
This string literal seems to be too big, perhaps consider labelling it as a named constant so that reference to its context is easier to be understood by readers of your code.

theopin reviewed (commented):
Good implementation on Duke so far! Some general pointers to improve your code:
- Use of magic literals: Try to reduce them by replacing them with named constants to enhance readability
- Long Methods: Seems like some methods have sub behaviour that can be abstracted to make its flow easier to be understood by readers

Commit:
Refeactor tasklist code

Commit:
Refector main file

Commit:
Add common messages

Commit:
Ad commands

Commit:
Add text ui

Commit:
Add parser

Commit:
Refecotr tasks and tasklist

Commit:
Align respsonse messages

Commit:
Ignore ctags files

Commit:
Add Deadline command

Commit:
Add event class

Commit:
Add mak/unmark command

Commit:
Add delete command

Commit:
Add storage

Commit:
clean up

Commit:
Merge branch 'branch-more-oop'

Commit:
Change date/time from String to LocalDate

imdxz committed:
Merge pull request #1 from danxuZhang/branch-Level-8

Level 8: Change date/time from String to LocalDate

Commit:
Add Level-9

Add find command to search for a keyword in the task list.

imdxz committed:
Merge pull request #2 from danxuZhang/branch-Level-9

Add Level-9

Commit:
Add Java Doc

Commit:
Add User Guide v0.2

imdxz committed:
Merge pull request #3 from danxuZhang/branch-A-JavaDoc

Add Java Doc

imdxz committed:
Merge pull request #4 from danxuZhang/branch-UG

Add User Guide v0.2

Commit:
Fix typos

Commit:
Add finally block
